File: User/lib/network/response_models/promo_model.dart
class PromoModel {
  String? zoneId;
  String? promoCode;
  String? promoCodeType;
  String? description;
  int? targetAmount;
  String? promoType;
  int? amount;
  int? percentage;
  String? clientId;
  List<dynamic>? userId;
  String? fromDate;
  String? toDate;
  bool? status;
  int? totalCount;
  String? banner;
  int? promoReuseCount;
  int? newUserCount;
  dynamic createdBy;
  String? id;

  PromoModel(
      {this.zoneId,
      this.promoCode,
      this.promoCodeType,
      this.description,
      this.targetAmount,
      this.promoType,
      this.amount,
      this.percentage,
      this.clientId,
      this.userId,
      this.fromDate,
      this.toDate,
      this.status,
      this.totalCount,
      this.banner,
      this.promoReuseCount,
      this.newUserCount,
      this.createdBy,
      this.id});

  PromoModel.fromJson(Map<String, dynamic> json) {
    zoneId = json['zoneId'];
    promoCode = json['promoCode'];
    promoCodeType = json['promoCodeType'];
    description = json['description'];
    targetAmount = json['targetAmount'];
    promoType = json['promoType'];
    amount = json['amount'];
    percentage = json['percentage'];
    clientId = json['clientId'];
    if (json['userId'] != null) {
      userId = <dynamic>[];
      json['userId'].forEach((v) {
        userId!.add(v);
      });
    }
    fromDate = json['fromDate'];
    toDate = json['toDate'];
    status = json['status'];
    totalCount = json['totalCount'];
    banner = json['banner'];
    promoReuseCount = json['promoReuseCount'];
    newUserCount = json['newUserCount'];
    createdBy = json['createdBy'];
    id = json['id'];
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = <String, dynamic>{};
    data['zoneId'] = this.zoneId;
    data['promoCode'] = this.promoCode;
    data['promoCodeType'] = this.promoCodeType;
    data['description'] = this.description;
    data['targetAmount'] = this.targetAmount;
    data['promoType'] = this.promoType;
    data['amount'] = this.amount;
    data['percentage'] = this.percentage;
    data['clientId'] = this.clientId;
    if (this.userId != null) {
      data['userId'] = this.userId!.map((v) => v).toList();
    }
    data['fromDate'] = this.fromDate;
    data['toDate'] = this.toDate;
    data['status'] = this.status;
    data['totalCount'] = this.totalCount;
    data['banner'] = this.banner;
    data['promoReuseCount'] = this.promoReuseCount;
    data['newUserCount'] = this.newUserCount;
    data['createdBy'] = this.createdBy;
    data['id'] = this.id;
    return data;
  }
}

File: Backend/src/models/boilerplate/promo.model.js
const mongoose = require('mongoose');
const { toJSON, paginate } = require('../plugins');

const promoCodeSchema = mongoose.Schema(
  {
    zoneId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Zone',
      required: true,
      index: true,
    },
    promoCode: {
      type: String,
      required: true,
      trim: true,
    },
    promoCodeType: {
      type: String,
      required: true,
      trim: true,
    },
    description: {
      type: String,
      required: true,
    },
    targetAmount: {
      type: Number,
      default: 0,
    },
    promoType: {
      type: String,
      default: null,
    },
    amount: {
      type: Number,
      default: 0,
    },
    percentage: {
      type: Number,
      default: 0,
    },
    clientId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Client',
    },
    userId: [
      {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
      },
    ],
    fromDate: {
      type: Date,
      default: null,
    },
    toDate: {
      type: Date,
      default: null,
    },
    status: {
      type: Boolean,
      required: true,
      default: true,
    },
    totalCount: {
      type: Number,
      required: true,
      default: 0,
    },
    banner: {
      type: String,
      required: false,
    },
    promoReuseCount: {
      type: Number,
      required: true,
      default: 0,
    },
    createdBy: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      default: null,
    },
    vehicleType: [{
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Vehicle',
    }],

  },
  {
    timestamps: true,
  },
);

promoCodeSchema.plugin(toJSON);
promoCodeSchema.plugin(paginate);

/**
 * @typedef PromoCode
 */
const PromoCode = mongoose.model('PromoCode', promoCodeSchema);

module.exports = PromoCode;
